## $\text{Written by EG Timerise aka Dr.E}$

$\text{ A simple introduction review of some basic R coding.}$

$\text{ Data set pulled from kaggle.com}$

$\text{An IBM sample data set}$

$\text{https://www.kaggle.com/blastchar/telco-customer-churn}$

$\text{Objective : Analyze customer data and develop customer focused retention plans.}$

$\text{First lets load the library 'readr' used to import our csv file.}$


$\text{A special note:}$

$\text{Note we are dealing with an observational data set and not a experimental data set.}$

$\text{We can use 'install.packages("tidyverse") ' to install the following libraries}$

$\text{'ggplot2' , 'tibble' , 'tidyr' , 'readr', 'purr','dplyr' , 'stringr'  , and 'forcats',etc...}$

$\text{We still need to load our desire library if we want to use it!}$


$\text{Some background information about the various variables:}$

#### customerID - Customer ID

#### gender - Customer gender (female, male)

#### SeniorCitizen - Whether the customer is a senior citizen or not (1, 0)

#### Partner -Whether the customer has a partner or not (Yes, No)

#### Dependents - Whether the customer has dependents or not (Yes, No)

#### tenure - Number of months the customer has stayed with the company

#### PhoneService - Whether the customer has a phone service or not (Yes, No)

#### MultipleLines - Whether the customer has multiple lines or not (Yes, No, No phone service)

#### InternetService - Customer’s internet service provider (DSL, Fiber optic, No)

#### OnlineSecurity - Whether the customer has online security or not (Yes, No, No internet service)

#### OnlineBackup - Whether the customer has online backup or not (Yes, No, No internet service)

#### DeviceProtection - Whether the customer has device protection or not (Yes, No, No internet service)

#### TechSupport - Whether the customer has tech support or not (Yes, No, No internet service)

#### StreamingTV - Whether the customer has streaming TV or not (Yes, No, No internet service)

#### StreamingMovies -Whether the customer has streaming movies or not (Yes, No, No internet service)

#### Contract - The contract term of the customer (Month-to-month, One year, Two year)

#### PaperlessBilling - Whether the customer has paperless billing or not (Yes, No)

#### PaymentMethod -The customer’s payment method (Electronic check, Mailed check, Bank transfer (automatic), Credit card (automatic))

#### MonthlyCharges - The amount charged to the customer monthly

#### TotalCharges -The total amount charged to the customer

#### Churn - Whether the customer churned or not (Yes or No), stop using service vs. conti. service

In [11]:
library(readr) ##loading the readr library 
#now importing our data set into the system
data1<- read_csv("C:/Users/dragon/Desktop/WA_Fn-UseC_-Telco-Customer-Churn.csv")

Parsed with column specification:
cols(
  .default = col_character(),
  SeniorCitizen = col_integer(),
  tenure = col_integer(),
  MonthlyCharges = col_double(),
  TotalCharges = col_double()
)
See spec(...) for full column specifications.


In [12]:
dim(data1) ## lets you view the dimension of our data set
## row vs columns, much like pythons 'data1.shape' format

[1] 7043   21

$\text{So it looks like we have 7043 observations and 21 variables.}$

$\text{Lets check out the first few observations.}$

In [13]:
head(data1,n=10) ## checking out the first 10 observations of our data 

customerID gender SeniorCitizen Partner Dependents tenure PhoneService
1  7590-VHVEG Female 0             Yes     No          1     No          
2  5575-GNVDE Male   0             No      No         34     Yes         
3  3668-QPYBK Male   0             No      No          2     Yes         
4  7795-CFOCW Male   0             No      No         45     No          
5  9237-HQITU Female 0             No      No          2     Yes         
6  9305-CDSKC Female 0             No      No          8     Yes         
7  1452-KIOVK Male   0             No      Yes        22     Yes         
8  6713-OKOMC Female 0             No      No         10     No          
9  7892-POOKP Female 0             Yes     No         28     Yes         
10 6388-TABGU Male   0             No      Yes        62     Yes         
   MultipleLines    InternetService OnlineSecurity ... DeviceProtection
1  No phone service DSL             No             ... No              
2  No               DSL             Yes            ... Yes             
3  No               DSL             Yes            ... No              
4  No phone service DSL             Yes            ... Yes             
5  No               Fiber optic     No             ... No              
6  Yes              Fiber optic     No             ... Yes             
7  Yes              Fiber optic     No             ... No              
8  No phone service DSL             Yes            ... No              
9  Yes              Fiber optic     No             ... Yes             
10 No               DSL             Yes            ... No              
   TechSupport StreamingTV StreamingMovies Contract       PaperlessBilling
1  No          No          No              Month-to-month Yes             
2  No          No          No              One year       No              
3  No          No          No              Month-to-month Yes             
4  Yes         No          No              One year       No              
5  No          No          No              Month-to-month Yes             
6  No          Yes         Yes             Month-to-month Yes             
7  No          Yes         No              Month-to-month Yes             
8  No          No          No              Month-to-month No              
9  Yes         Yes         Yes             Month-to-month Yes             
10 No          No          No              One year       No              
   PaymentMethod             MonthlyCharges TotalCharges Churn
1  Electronic check           29.85           29.85      No   
2  Mailed check               56.95         1889.50      No   
3  Mailed check               53.85          108.15      Yes  
4  Bank transfer (automatic)  42.30         1840.75      No   
5  Electronic check           70.70          151.65      Yes  
6  Electronic check           99.65          820.50      Yes  
7  Credit card (automatic)    89.10         1949.40      No   
8  Mailed check               29.75          301.90      No   
9  Electronic check          104.80         3046.05      Yes  
10 Bank transfer (automatic)  56.15         3487.95      No

$\text{From the above code, we can see the general layout of a few of the 21 variables.}$

$\text{Our general target variable, Churn, was placed at the end of the data set. This is a bit different from the standard method.}$

$\text{Generally the target variable is placed at the start of the data set, but it doesn't really matter.}$

$\text{Lets change the data set to fix our desire layout.}$


In [15]:
library(dplyr) ## a library useful for transforming data frames
##check out the following pdf written 2018, for more about the library
#https://cran.r-project.org/web/packages/dplyr/dplyr.pdf

### Applying a chain method of coding inplace of using (())
## Think of %>% as the word then, check it in action below!
data1=data1 %>%
    select(Churn,everything(),-c(customerID))  ## switch the order and droping ID

data1=data1 %>% 
    mutate_if(is.character,as.factor) ##changing chr types to factor types

In [16]:
head(data1)

Churn gender SeniorCitizen Partner Dependents tenure PhoneService
1 No    Female 0             Yes     No          1     No          
2 No    Male   0             No      No         34     Yes         
3 Yes   Male   0             No      No          2     Yes         
4 No    Male   0             No      No         45     No          
5 Yes   Female 0             No      No          2     Yes         
6 Yes   Female 0             No      No          8     Yes         
  MultipleLines    InternetService OnlineSecurity OnlineBackup DeviceProtection
1 No phone service DSL             No             Yes          No              
2 No               DSL             Yes            No           Yes             
3 No               DSL             Yes            Yes          No              
4 No phone service DSL             Yes            No           Yes             
5 No               Fiber optic     No             No           No              
6 Yes              Fiber optic     No             No           Yes             
  TechSupport StreamingTV StreamingMovies Contract       PaperlessBilling
1 No          No          No              Month-to-month Yes             
2 No          No          No              One year       No              
3 No          No          No              Month-to-month Yes             
4 Yes         No          No              One year       No              
5 No          No          No              Month-to-month Yes             
6 No          Yes         Yes             Month-to-month Yes             
  PaymentMethod             MonthlyCharges TotalCharges
1 Electronic check          29.85            29.85     
2 Mailed check              56.95          1889.50     
3 Mailed check              53.85           108.15     
4 Bank transfer (automatic) 42.30          1840.75     
5 Electronic check          70.70           151.65     
6 Electronic check          99.65           820.50

$\text{Notice that I removed the customerID, for the type of analysis I am  conducting the ID numbers are not needed.}$

$\text{I also turned the chr type variables into factor types.}$

$\text{We want to review  the data to come up with a good retention plan.}$

$\text{Lets dive deeper into the data to develope some simple introductory analysis!}$

In [17]:
summary(data1) # summary() gives us quick single variable information

 Churn         gender     SeniorCitizen    Partner    Dependents
 No :5174   Female:3488   Min.   :0.0000   No :3641   No :4933  
 Yes:1869   Male  :3555   1st Qu.:0.0000   Yes:3402   Yes:2110  
                          Median :0.0000                        
                          Mean   :0.1621                        
                          3rd Qu.:0.0000                        
                          Max.   :1.0000                        
                                                                
     tenure      PhoneService          MultipleLines     InternetService
 Min.   : 0.00   No : 682     No              :3390   DSL        :2421  
 1st Qu.: 9.00   Yes:6361     No phone service: 682   Fiber optic:3096  
 Median :29.00                Yes             :2971   No         :1526  
 Mean   :32.37                                                          
 3rd Qu.:55.00                                                          
 Max.   :72.00                            

$\text{From the above out put , we can see that our data set is mix. (also from the pre-information given.}$

$\text{From TotalCharges, we see that we have some NA's to fix.}$

$\text{Lets check out TotalCharges in greater detail.}$

In [18]:
head(sort(data1$TotalCharges))## sort by default starts from lowest value than biggest
#Although we already knew 18.8 was the lowest written value

[1] 18.80 18.85 18.85 18.90 19.00 19.05

In [19]:
head(sort(data1$tenure),n=13)## sort by default starts from lowest value than biggest


[1] 0 0 0 0 0 0 0 0 0 0 0 1 1

$\text{To be  truthful, we really do not know what the nans in TotalCharges were suppose to be.}$

$\text{We could replace the values with MonthlyCharges+(MonthlyCharges*tenure) , or etc...,or etc..}$

$\text{It seems likely that the 11 observations with NAs for TotalCharges are those with less than a month of use.}$

$\text{We could just remove the observations, since its only 11 out of 7043, but lets check out the weird observations first.}$

In [20]:
data1[rowSums(is.na(data1))>0,]## using boolean logic if any row has an NA 'True/1' then it will print

## df[row,column] style format much like pythons 'df.iloc[row,column] ' kind of



Churn gender SeniorCitizen Partner Dependents tenure PhoneService
1  No    Female 0             Yes     Yes        0      No          
2  No    Male   0             No      Yes        0      Yes         
3  No    Female 0             Yes     Yes        0      Yes         
4  No    Male   0             Yes     Yes        0      Yes         
5  No    Female 0             Yes     Yes        0      No          
6  No    Male   0             Yes     Yes        0      Yes         
7  No    Male   0             Yes     Yes        0      Yes         
8  No    Female 0             Yes     Yes        0      Yes         
9  No    Male   0             Yes     Yes        0      Yes         
10 No    Female 0             Yes     Yes        0      Yes         
11 No    Male   0             No      Yes        0      Yes         
   MultipleLines    InternetService OnlineSecurity      OnlineBackup       
1  No phone service DSL             Yes                 No                 
2  No               No              No internet service No internet service
3  No               DSL             Yes                 Yes                
4  Yes              No              No internet service No internet service
5  No phone service DSL             Yes                 Yes                
6  No               No              No internet service No internet service
7  Yes              No              No internet service No internet service
8  No               No              No internet service No internet service
9  No               No              No internet service No internet service
10 Yes              DSL             No                  Yes                
11 Yes              DSL             Yes                 Yes                
   DeviceProtection    TechSupport         StreamingTV        
1  Yes                 Yes                 Yes                
2  No internet service No internet service No internet service
3  Yes                 No                  Yes                
4  No internet service No internet service No internet service
5  Yes                 Yes                 Yes                
6  No internet service No internet service No internet service
7  No internet service No internet service No internet service
8  No internet service No internet service No internet service
9  No internet service No internet service No internet service
10 Yes                 Yes                 Yes                
11 No                  Yes                 No                 
   StreamingMovies     Contract PaperlessBilling PaymentMethod            
1  No                  Two year Yes              Bank transfer (automatic)
2  No internet service Two year No               Mailed check             
3  Yes                 Two year No               Mailed check             
4  No internet service Two year No               Mailed check             
5  No                  Two year No               Credit card (automatic)  
6  No internet service Two year No               Mailed check             
7  No internet service Two year No               Mailed check             
8  No internet service Two year No               Mailed check             
9  No internet service One year Yes              Mailed check             
10 No                  Two year No               Mailed check             
11 No                  Two year Yes              Bank transfer (automatic)
   MonthlyCharges TotalCharges
1  52.55          NA          
2  20.25          NA          
3  80.85          NA          
4  25.75          NA          
5  56.05          NA          
6  19.85          NA          
7  25.35          NA          
8  20.00          NA          
9  19.70          NA          
10 73.35          NA          
11 61.90          NA

$\text{As I suspected, the observations with NAs for TotalCharges are accounts with less than one month.}$

$\text{It appears that charges are only made or counted at the end of a month.}$

$\text{To stay with the input logic, we can set our NA values to 0.}$

In [21]:
data1$TotalCharges[is.na(data1$TotalCharges)] <-0  ## This is replacing the na's in totalcharges with zeros
## a bit different from python, where you need to save your changes in a new object. 'data1.totalcharges.replace(np.nan,0,inplace=True)'

$\text{Now lets check for duplicate observations in our data set!}$

In [22]:
dim(data1)

[1] 7043   20

In [23]:
library(dplyr)

In [24]:
dim(distinct(data1)) ## removes duplicate rows based on each columns

[1] 7021   20

$\text{There does not appear to be any duplicate observations in the data set.}$

$\text{With python , we could have used 'data1.duplicated().sum()' to check for duplicate rows.}$

$\text{Now lets check out the various types of variables we are dealing with in R.}$

In [25]:
str(data1) ## shows the deeper information about the data set
## Its much like pythons  'data1.info() ' 

## In R could also use the class function 'class(data1$gender)'
## In python it would be 'type(data1.gender)'

Classes 'tbl_df', 'tbl' and 'data.frame':	7043 obs. of  20 variables:
 $ Churn           : Factor w/ 2 levels "No","Yes": 1 1 2 1 2 2 1 1 2 1 ...
 $ gender          : Factor w/ 2 levels "Female","Male": 1 2 2 2 1 1 2 1 1 2 ...
 $ SeniorCitizen   : int  0 0 0 0 0 0 0 0 0 0 ...
 $ Partner         : Factor w/ 2 levels "No","Yes": 2 1 1 1 1 1 1 1 2 1 ...
 $ Dependents      : Factor w/ 2 levels "No","Yes": 1 1 1 1 1 1 2 1 1 2 ...
 $ tenure          : int  1 34 2 45 2 8 22 10 28 62 ...
 $ PhoneService    : Factor w/ 2 levels "No","Yes": 1 2 2 1 2 2 2 1 2 2 ...
 $ MultipleLines   : Factor w/ 3 levels "No","No phone service",..: 2 1 1 2 1 3 3 2 3 1 ...
 $ InternetService : Factor w/ 3 levels "DSL","Fiber optic",..: 1 1 1 1 2 2 2 1 2 1 ...
 $ OnlineSecurity  : Factor w/ 3 levels "No","No internet service",..: 1 3 3 3 1 1 1 3 1 3 ...
 $ OnlineBackup    : Factor w/ 3 levels "No","No internet service",..: 3 1 3 1 1 1 3 1 1 3 ...
 $ DeviceProtection: Factor w/ 3 levels "No","No internet service",..

$\text{I see that the variables are listed as Factor, num, or int type.}$